File: solvers/differentiation.ml
open Core

open Type
open Program
open Utils
    
type variable = {mutable gradient : float option;
                 mutable data : float option;
                 volatile : bool; (* whether this depends on something which is a function of learned parameters *)
                 arguments : variable list;
                 forwardProcedure : float list -> float;
                 backwardProcedure : float list -> float list;
                 (* descendents: every variable that takes this variable as input *)
                 (* the additional float parameter is d Descendent / d This *)
                 mutable descendents : (variable*float) list}

let variable_value v = v.data |> get_some

let make_variable forward backward arguments =
  let values = List.map ~f:(fun v -> v.data |> get_some) arguments in
  let initial_value = forward values in
  let s = {gradient = None;
           arguments = arguments;
           volatile = List.exists arguments ~f:(fun a -> a.volatile);
           descendents = [];
           forwardProcedure = forward;
           backwardProcedure = backward;
           data = Some(initial_value);}
  in
  let initial_partials = backward values in
  List.iter2_exn initial_partials arguments ~f:(fun p a ->
      if a.volatile then
        a.descendents <- (s,p)::a.descendents);
  s

let make_binary_variable forward backward a b =
  make_variable (fun x -> match x with
      |[a;b] -> forward a b
      |_ -> raise (Failure "Binary variable did not get 2 arguments"))
    (fun x -> match x with
      |[a;b] -> backward a b
      |_ -> raise (Failure "Binary variable did not get 2 backward arguments"))
    [a;b;]

let make_unitary_variable forward backward a =
  make_variable (fun x -> match x with
      |[a;] -> forward a 
      |_ -> raise (Failure "unitary variable did not get 1 arguments"))
    (fun x -> match x with
      |[a;] -> backward a 
      |_ -> raise (Failure "urinary variable did not get 1 backward arguments"))
    [a;]

let (~$) x =
  let rec s = {gradient = None;
               arguments = [];
               volatile = false;
               descendents = [];
               data = Some(x);
               forwardProcedure = (fun _ -> s.data |> get_some);
               backwardProcedure = (fun _ -> []);
              }
  in s

let placeholder () =
  let rec s = {gradient = None;
               arguments = [];
               descendents = [];
               volatile = true;
               data = None;
               forwardProcedure = (fun _ -> s.data |> get_some);
               backwardProcedure = (fun _ -> []);
              }
  in s

let random_variable ?mean:(mean = 0.) ?standard_deviation:(standard_deviation = 1.) () =
  let rec s = {gradient = None;
               arguments = [];
               descendents = [];
               volatile = true;
               data = Some(normal standard_deviation mean);
               forwardProcedure = (fun _ -> s.data |> get_some);
               backwardProcedure = (fun _ -> []);
              }
  in s

let update_variable v x =
  assert (v.arguments = []);
  v.data <- Some(x)


let (+&) =
  make_binary_variable (+.) (fun _ _ -> [1.;1.]) 

let (-&) =
  make_binary_variable (-.) (fun _ _ -> [1.;-1.]) 

let ( *& ) =
  make_binary_variable ( *. ) (fun a b -> [b;a])

let ( /& ) =
  make_binary_variable ( /. ) (fun a b -> [1./.b;0.-.a/.(b*.b)])

let power =
  make_binary_variable ( ** ) (fun a b -> [b*.(a**(b-.1.));
                                          (a**b) *. (log a)])

let logarithm =
  make_unitary_variable log (fun a -> [1./.a])

let exponential =
  make_unitary_variable exp (fun a -> [exp a])

let square =
  make_unitary_variable (fun a -> a*.a) (fun a -> [2.*.a])

let square_root =
  make_unitary_variable sqrt (fun a -> [0.5/.(sqrt a)])

let clamp ~l ~u =
  make_unitary_variable (fun a ->
      if a > u then u else
      if a < l then l else
        a)
    (fun a ->
       if a > u || a < l then [0.] else [1.])
    

let log_soft_max xs =
  make_variable (fun vs -> 
      let m : float = List.fold_right vs ~init:Float.neg_infinity  ~f:max in
      let zm = List.fold_right ~init:0. ~f:(fun x a -> exp (x -. m) +. a) vs in
      m+. (log zm))
    (fun vs -> 
      let m : float = List.fold_right vs ~init:Float.neg_infinity  ~f:max in
      let zm = List.fold_right ~init:0. ~f:(fun x a -> exp (x -. m) +. a) vs in
      List.map vs ~f:(fun x -> (exp (x-.m)) /. zm))
    xs


let rec differentiate z =
  match z.gradient with
  | Some(g) -> g
  | None ->
    let g = List.fold_right z.descendents  ~init:0.0 ~f:(fun (d,partial) a -> a +. differentiate d *. partial) in
    z.gradient <- Some(g);
    g

let rec zero_gradients z =
  match z.gradient with
  | None -> ()
  | Some(_) -> begin
      List.iter z.arguments ~f:(fun a -> zero_gradients a);
      z.gradient <- None;
      z.descendents <- [];
      if z.arguments = [] then () else z.data <- None
    end

let rec forward z =
  match z.data with
  | Some(v) -> v
  | None ->
    let argument_values = List.map z.arguments ~f:forward in
    let v = z.forwardProcedure argument_values in
    let partials = z.backwardProcedure argument_values in
    List.iter2_exn partials z.arguments ~f:(fun p a ->
        a.descendents <- (z,p)::a.descendents);
    z.data <- Some(v);
    v

let backward z =
  z.gradient <- Some(1.0);
  let rec b v =
    ignore(differentiate v);
    List.iter ~f:b v.arguments
  in b z

let update_network loss =
  zero_gradients loss;
  let l = forward loss in
  backward loss;
  l

let rec run_optimizer opt ?update:(update = 1000)
    ?iterations:(iterations = 10000) parameters loss =
  let l = update_network loss in
  if iterations = 0 then l else begin

    if update > 0 && iterations mod update = 0 then begin 
      Printf.eprintf "LOSS: %f\n" l;
      parameters |> List.iter ~f:(fun p -> Printf.eprintf "parameter %f\t" (p.data |> get_some));
      Printf.eprintf "\n";
    end else ();

    parameters |> List.map ~f:differentiate |> opt |> 
    List.iter2_exn parameters ~f:(fun x dx ->
        let v = x.data |> get_some in
        update_variable x (v +. dx));
    run_optimizer opt ~update:update ~iterations:(iterations - 1) parameters loss
  end

let restarting_optimize opt ?update:(update = 1000)
    ?attempts:(attempts=1)
    ?iterations:(iterations = 10000) parameters loss =
  (0--attempts) |> List.map ~f:(fun _ ->
      parameters |> List.iter ~f:(fun parameter ->
          update_variable parameter (uniform_interval ~l:(-5.) ~u:5.));
      run_optimizer opt ~update:update ~iterations:iterations parameters loss) |>
  fold1 min

let gradient_descent ?lr:(lr = 0.001) =
  List.map ~f:(fun dx -> ~-. (lr*.dx))

let rprop ?lr:(lr=0.1) ?decay:(decay=0.5) ?grow:(grow=1.2) = 
  let first_iteration = ref true in
  let previous_signs = ref [] in
  let individual_rates = ref [] in

  fun dxs ->
    let new_signs = dxs |> List.map ~f:(fun dx -> dx > 0.) in
    if !first_iteration then begin
      first_iteration := false;
      (* First iteration: ignore the previous signs, which have not yet been recorded *)
      let updates = dxs |> List.map ~f:(fun dx -> ~-. dx*.lr) in

      previous_signs := new_signs;
      individual_rates := dxs |> List.map ~f:(fun _ -> lr);

      updates      
    end else begin 
      individual_rates := List.map3_exn !individual_rates !previous_signs new_signs
          ~f:(fun individual_rate previous_sign new_sign ->
              if previous_sign = new_sign
              then individual_rate*.grow
              else individual_rate*.decay);
      
      let updates = List.map2_exn !individual_rates dxs
          ~f:(fun individual_rate dx ->
              if dx > 0.
              then ~-. individual_rate
              else if dx < 0. then individual_rate else 0.)
      in
      previous_signs := new_signs;
      updates
    end


let proportional_optimize parameters loss_pairs =
  parameters |> List.iter ~f:(fun p -> update_variable p 1.);
  let prediction_variables = loss_pairs |> List.map ~f:fst in
  prediction_variables |> List.iter ~f:zero_gradients;
  let predictions = prediction_variables |> List.map ~f:forward in
  assert (false) (* TODO *)

let test_differentiation () =
  let x = ~$ 10.0 in
  let y = ~$ 2. in
  let z = x -& log_soft_max [x;y;] in
  backward z;
  Printf.printf "dL/dx = %f\tdL/dy = %f\n" (differentiate x) (differentiate y);

  update_variable x 2.;
  update_variable y 10.;

  ignore(update_network z);

  Printf.printf "dL/dx = %f\tdL/dy = %f\n" (differentiate x) (differentiate y);  

  update_variable x 2.;
  update_variable y 2.;

  ignore(update_network z);

  Printf.printf "z = %f\n" (z.data |> get_some);

  Printf.printf "dL/dx = %f\tdL/dy = %f\n" (differentiate x) (differentiate y);

  let l = ((~$ 0.) -& z) in
  ignore(run_optimizer (gradient_descent ~lr:0.001) [x;y] l)

;;

(* Integration with programs *)
let differentiable_zero = primitive "0." treal (~$ 0.);;
let differentiable_one = primitive "1." treal (~$ 1.);;
let differentiable_pi = primitive "pi" treal (~$ 3.14);;
let differentiable_add = primitive "+." (treal @> treal @> treal) (+&);;
let differentiable_subtract = primitive "-." (treal @> treal @> treal) (-&);;
let differentiable_multiply = primitive "*." (treal @> treal @> treal) ( *&);;
let differentiable_division = primitive "/." (treal @> treal @> treal) ( /&);;
let differentiable_power = primitive "power" (treal @> treal @> treal) (power);;
let differentiable_placeholder = primitive "REAL" treal ();;

let replace_placeholders program =
  let placeholders = ref [] in
  let rec r = function
    | Index(j) -> Index(j)
    | Abstraction(b) -> Abstraction(r b)
    | Apply(f,x) -> Apply(r f, r x)
    | Invented(t,b) -> Invented(t,r b)
    | Primitive(t,"REAL",_) -> begin
        let v = random_variable() in
        (* update_variable v 0.; *)
        placeholders := v :: !placeholders;
        Primitive(t,"REAL", ref v |> magical)
      end
    | Primitive(t,"real",v') -> begin
        let v = random_variable() in
        update_variable v (!(magical v'));
        (* update_variable v 0.; *)
        placeholders := v :: !placeholders;
        Primitive(t,"REAL", ref v |> magical)
      end
    | p -> p
  in
  let program = r program in
  (program, !placeholders)

let rec placeholder_data t x =
  match t with
  | TCon("real",_,_) -> magical (~$ (magical x))
  | TCon("list",[tp],_) ->
    magical x |> List.map ~f:(placeholder_data tp) |> magical
  | _ -> raise (Failure ("placeholder_data: bad type "^(string_of_type t)))


exception DifferentiableBadShape

(* Takes as input a type *)
(* Returns a function from (prediction, target) to [(prediction_variable, target_variable)] *)
let rec polymorphic_loss_pairs : tp -> 'a -> 'b -> (variable*variable) list = function
  | TCon("vector",_,_) -> polymorphic_loss_pairs (tlist treal)
  | TCon("real",_,_) -> magical (fun p y -> [(p,y)])
  | TCon("list",[tp],_) ->
    let e = polymorphic_loss_pairs tp in
    magical (fun p y ->
        try
          List.fold2_exn p y ~init:[] ~f:(fun a _p _y -> a @ (e _p _y))
        with _ -> raise DifferentiableBadShape)
  | t -> raise (Failure ("placeholder_data: bad type "^(string_of_type t)))


let rec polymorphic_sse ?clipOutput:(clipOutput=None) ?clipLoss:(clipLoss=None) t =
  let get_pairs = polymorphic_loss_pairs t in
  fun p y ->
    let loss_pairs = get_pairs p y in
    loss_pairs |> List.map ~f:(fun (p,y) ->
        let p = match clipOutput with
          | None -> p
          | Some(clip) -> clamp ~l:(-.clip) ~u:clip p
        in
        let l = square (p -& y) in
        match clipLoss with
        | None -> l
        | Some(clip) -> clamp ~l:0. ~u:clip l) |>
    List.reduce_exn ~f:(+&)
(* let rec polymorphic_sse ?clipOutput:(clipOutput=None) ?clipLoss:(clipLoss=None) = function *)
(*   | TCon("vector",_,_) -> polymorphic_sse ~clipOutput ~clipLoss (tlist treal) *)
(*   | TCon("real",_,_) -> magical (fun p y -> *)
(*       let p = match clipOutput with *)
(*         | None -> p *)
(*         | Some(clip) -> clamp ~l:(-.clip) ~u:clip p *)
(*       in *)
(*       let l = square (p -& y) in *)
(*       match clipLoss with *)
(*       | None -> l *)
(*       | Some(clip) -> clamp ~l:0. ~u:clip l) *)
(*   | TCon("list",[tp],_) -> *)
(*     let e = polymorphic_sse ~clipOutput ~clipLoss tp in *)
(*     magical (fun p y -> *)
(*         try *)
(*           List.fold2_exn p y ~init:(~$0.) ~f:(fun a _p _y -> a +& (e _p _y)) *)
(*         with _ -> raise DifferentiableBadShape) *)
(*   | t -> raise (Failure ("placeholder_data: bad type "^(string_of_type t))) *)
  


let test_program_differentiation() =
  let p = parse_program "(lambda REAL)" |> get_some in
  let (p, parameters) = replace_placeholders p in
  let p = analyze_lazy_evaluation p in

  let g = run_lazy_analyzed_with_arguments p [~$ 0.] in

  Printf.printf "%f" (update_network g);;




(* test_differentiation();; *)

(* type symbolic_expression = *)
(*   | SymbolicUnit | SymbolicZero *)
(*   | SymbolicVariable *)
(*   | SymbolicConstant of int *)
(*   | Summation of symbolic_expression list *)
(*   | Product of symbolic_expression list *)
(*   | Quotient of symbolic_expression*symbolic_expression *)
(* [@@deriving compare] *)

(* let make_summation x y = match x,y with *)
(*   | SymbolicZero,a | a,SymbolicZero -> a *)
(*   | _ -> Summation([x;y;]);; *)
(* let make_product x y = match x,y with *)
(*   | SymbolicUnit,a|a,SymbolicUnit -> a *)
(*   | SymbolicZero,_|_,SymbolicZero -> SymbolicZero *)
(*   | _ -> Product([x;y;]) *)

(* let make_quotient p q = match p,q with *)
(*   | _,SymbolicUnit -> p *)
(*   | _,_ -> Quotient(p,q) *)

(* let rec show_symbolic = function *)
(*   | SymbolicUnit -> "1" | SymbolicZero -> "0" | SymbolicVariable -> "x" *)
(*   | SymbolicConstant(j) -> Printf.sprintf "r%d" j *)
(*   | Summation(ss) -> ss |> List.map ~f:show_symbolic |> join ~separator:" + " |> Printf.sprintf "(%s)" *)
(*   | Product(ss) -> ss |> List.map ~f:show_symbolic |> join ~separator:" * " |> Printf.sprintf "(%s)" *)
(*   | Quotient(n,d) -> Printf.sprintf "(%s / %s)" (show_symbolic n) (show_symbolic d) *)
    


(* let program_to_symbolic_expression p = *)
(*   let counter = ref 0 in *)
(*   let rec annotate = function *)
(*     | Primitive(t,"REAL",_) -> (incr counter; Primitive(t,"REAL",!counter |> ref  |> magical)) *)
(*     | Apply(f,x) -> Apply(annotate f, *)
(*                           annotate x) *)
(*     | Abstraction(b) -> Abstraction(annotate b) *)
(*     | Invented(t,b) -> Invented(t,b |> annotate) *)
(*     | e -> e *)
(*   in  *)
(*   let rec walk = function *)
(*     | Primitive(_,"REAL",c) -> SymbolicConstant(!(c |> magical)) *)
(*     | Index(0) -> SymbolicVariable *)
(*     | Apply(Apply(Primitive(_,"+.",_),x),y) -> *)
(*       make_summation (walk x) (walk y)                  *)
(*     | Apply(Apply(Primitive(_,"*.",_),x),y) -> *)
(*       make_product *)
(*         (walk x) *)
(*         (walk y) *)
(*     | Apply(Apply(Primitive(_,"/.",_),x),y) -> *)
(*       Quotient(walk x, *)
(*                walk y) *)
(*     | Abstraction(b) -> walk b *)
(*     | _ -> assert (false) *)
(*   in *)
(*   annotate p |> beta_normal_form ~reduceInventions:true |> walk *)

(* type canonical_polynomial = symbolic_expression list *)

(* type canonical_expression = canonical_polynomial*canonical_polynomial *)


(* let show_polynomial p = *)
(*   p |> List.mapi ~f:(fun e c -> Printf.sprintf "%sx^%d" (show_symbolic c) e) |> *)
(*   join ~separator:" + ";; *)

(* let show_canonical (n,d) = *)
(*   Printf.sprintf "(%s) / (%s)" (show_polynomial n) (show_polynomial d) *)

(* let rec simplify_symbolic s : canonical_expression = *)
(*   let multiply_polynomial_by_x p = SymbolicZero :: p in *)

(*   let add_polynomials p q = *)
(*     let d = max (List.length p) (List.length q) in *)
(*     let p = p @ (List.range (List.length p) d |> List.map ~f:(fun _ -> SymbolicZero)) in *)
(*     let q = q @ (List.range (List.length q) d |> List.map ~f:(fun _ -> SymbolicZero)) in *)
(*     List.map2_exn p q ~f:make_summation *)
(*   in  *)
  
(*   let rec multiply_polynomials p q = *)
(*     match p with *)
(*     | [] -> [] *)
(*     | k :: p -> *)
(*       add_polynomials *)
(*         (q |> List.map ~f:(fun q' -> make_product k q')) *)
(*         (multiply_polynomials p q |> multiply_polynomial_by_x) *)
(*   in *)
  
(*   let rec simplify_product (components : canonical_polynomial list) = *)
(*     match components with *)
(*     | [] -> assert (false) *)
(*     | [c] -> c *)
(*     | x :: y :: z -> *)
(*       multiply_polynomials x y :: z |> simplify_product *)
(*   in  *)
    
(*   match s with *)
(*   | SymbolicVariable -> [SymbolicZero;SymbolicUnit;], [SymbolicUnit] *)
(*   | SymbolicConstant(_) -> [s], [SymbolicUnit] *)
(*   | Product(ss) -> *)
(*     let ss = ss |> List.map ~f:simplify_symbolic in *)
(*     let numerator = ss |> List.map ~f:fst |> simplify_product in *)
(*     let denominator = ss |> List.map ~f:snd |> simplify_product in *)
(*     (numerator, denominator) *)
(*   | Summation([x;y;]) -> *)
(*     let x = simplify_symbolic x in *)
(*     let y = simplify_symbolic y in *)
(*     let denominator = simplify_product [snd x;snd y] in *)
(*     let numerator = add_polynomials *)
(*         (multiply_polynomials (fst x) (snd y)) *)
(*         (multiply_polynomials (fst y) (snd x)) *)
(*     in *)
(*     numerator, denominator *)
(*   | Quotient(p,q) -> *)
(*     let n,d = simplify_symbolic p in *)
(*     let n',d' = simplify_symbolic q in *)
(*     (multiply_polynomials n d', *)
(*      multiply_polynomials d n') *)
(*   | _ -> assert (false) *)

(* let simplify_canonical_expression (n,d) = *)
(*   let leading_coefficient = List.last_exn d in *)
(*   let d = d |> List.map ~f:(fun d' -> make_quotient d' leading_coefficient) in *)
(*   let n = n |> List.map ~f:(fun d' -> make_quotient d' leading_coefficient) in *)

(*   let rec simplify_coefficient : symbolic_expression -> symbolic_expression = function *)
(*     | SymbolicVariable -> assert (false) *)
(*     | SymbolicUnit -> SymbolicUnit *)
(*     | SymbolicConstant(j) -> SymbolicConstant(j) *)
(*     | Summation(ss) -> *)
(*       let ss = ss |> List.map ~f:simplify_coefficient |> List.map ~f:(function *)
(*           | Summation(ss') -> ss' *)
(*           | SymbolicZero -> [] *)
(*           | z -> [z]) |> List.concat *)
(*       in *)
(*       Summation(ss |> List.sort ~compare:compare_symbolic_expression) *)
(*     | Product(ss) -> *)
(*       let ss = ss |> List.map ~f:simplify_coefficient in *)
(*       assert (false) *)
      

(*   (n |> List.map ~f:simplify_coefficient, *)
(*    d |> List.map ~f:simplify_coefficient) *)

(* let test_simplify() = *)
(*   let p = "(lambda (+. (/. (+. $0 (\*. $0 REAL)) REAL) REAL))" |> parse_program |> get_some in *)
(*   let s = program_to_symbolic_expression p in *)
(*   Printf.printf "%s\n" (show_symbolic s); *)
(*   let c = simplify_symbolic s in *)
(*   Printf.printf "%s\n" *)
(*     (show_canonical c) *)
(*   ; *)
(*   exit *)
(*   0 *)
(* ;; *)

(* test_simplify() *)
    

    
   

File: dreamcoder/differentiation.py
import math
import random
from dreamcoder.utilities import *


class InvalidLoss(Exception):
    pass


class DN(object):
    '''differentiable node: parent object of every differentiable operation'''

    def __init__(self, arguments):
        self.gradient = None
        if arguments != []:
            self.data = None
        self.arguments = arguments

        # descendents: every variable that takes this variable as input
        # descendents: [(DN,float)]
        # the additional float parameter is d Descendent / d This
        self.descendents = []

        self.recalculate()

    def __str__(self):
        if self.arguments == []:
            return self.name
        return "(%s %s)" % (self.name, " ".join(str(x)
                                                for x in self.arguments))

    def __repr__(self):
        return "DN(op = %s, data = %s, grad = %s, #descendents = %d, args = %s)" % (
            self.name, self.data, self.gradient, len(self.descendents), self.arguments)

    @property
    def derivative(self): return self.differentiate()

    def differentiate(self):
        if self.gradient is None:
            self.gradient = sum(partial * descendent.differentiate()
                                for descendent, partial in self.descendents)
        return self.gradient

    def zeroEverything(self):
        if self.gradient is None and self.descendents == [] and (
                self.data is None or self.arguments == []):
            return

        self.gradient = None
        self.descendents = []
        if self.arguments != []:
            self.data = None

        for x in self.arguments:
            x.zeroEverything()

    def lightweightRecalculate(self):
        return self.forward(*[a.lightweightRecalculate()
                              for a in self.arguments])

    def recalculate(self):
        if self.data is None:
            inputs = [a.recalculate() for a in self.arguments]
            self.data = self.forward(*inputs)
            # if invalid(self.data):
            #     eprint("I am invalid",repr(self))
            #     eprint("Here are my inputs",inputs)
            #     self.zeroEverything()
            #     eprint("Here I am after being zeroed",repr(self))
            #     raise Exception('invalid loss')
            #assert valid(self.data)
            partials = self.backward(*inputs)
            for d, a in zip(partials, self.arguments):
                # if invalid(d):
                #     eprint("I have an invalid derivative",self)
                #     eprint("Inputs",inputs)
                #     eprint("partials",partials)
                #     raise Exception('invalid derivative')
                a.descendents.append((self, d))
        return self.data

    def backPropagation(self):
        self.gradient = 1.
        self.recursivelyDifferentiate()

    def recursivelyDifferentiate(self):
        self.differentiate()
        for x in self.arguments:
            x.recursivelyDifferentiate()

    def updateNetwork(self):
        self.zeroEverything()
        l = self.recalculate()
        self.backPropagation()
        return l

    def log(self): return Logarithm(self)

    def square(self): return Square(self)

    def exp(self): return Exponentiation(self)

    def clamp(self, l, u): return Clamp(self, l, u)

    def __abs__(self): return AbsoluteValue(self)

    def __add__(self, o): return Addition(self, Placeholder.maybe(o))

    def __radd__(self, o): return Addition(self, Placeholder.maybe(o))

    def __sub__(self, o): return Subtraction(self, Placeholder.maybe(o))

    def __rsub__(self, o): return Subtraction(Placeholder.maybe(o), self)

    def __mul__(self, o): return Multiplication(self, Placeholder.maybe(o))

    def __rmul__(self, o): return Multiplication(self, Placeholder.maybe(o))

    def __neg__(self): return Negation(self)

    def __truediv__(self, o): return Division(self, Placeholder.maybe(o))

    def __rtruediv__(self, o): return Division(Placeholder.maybe(o), self)

    def numericallyVerifyGradients(self, parameters):
        calculatedGradients = [p.derivative for p in parameters]
        e = 0.00001
        for j, p in enumerate(parameters):
            p.data -= e
            y1 = self.lightweightRecalculate()
            p.data += 2 * e
            y2 = self.lightweightRecalculate()
            p.data -= e
            d = (y2 - y1) / (2 * e)
            if abs(calculatedGradients[j] - d) > 0.1:
                eprint(
                    "Bad gradient: expected %f, got %f" %
                    (d, calculatedGradients[j]))

    def gradientDescent(
            self,
            parameters,
            _=None,
            lr=0.001,
            steps=10**3,
            update=None):
        for j in range(steps):
            l = self.updateNetwork()
            if update is not None and j % update == 0:
                eprint("LOSS:", l)
                for p in parameters:
                    eprint(p.data, '\t', p.derivative)
            if invalid(l):
                raise InvalidLoss()

            for p in parameters:
                p.data -= lr * p.derivative
        return self.data

    def restartingOptimize(self, parameters, _=None, attempts=1,
                           s=1., decay=0.5, grow=0.1,
                           lr=0.1, steps=10**3, update=None):
        ls = []
        for _ in range(attempts):
            for p in parameters:
                p.data = random.random()*10 - 5
            ls.append(
                self.resilientBackPropagation(
                    parameters, lr=lr, steps=steps,
                    decay=decay, grow=grow))
        return min(ls)

    def resilientBackPropagation(
            self,
            parameters,
            _=None,
            decay=0.5,
            grow=1.2,
            lr=0.1,
            steps=10**3,
            update=None):
        previousSign = [None] * len(parameters)
        lr = [lr] * len(parameters)
        for j in range(steps):
            l = self.updateNetwork()

            if update is not None and j % update == 0:
                eprint("LOSS:", l)
                eprint("\t".join(str(p.derivative) for p in parameters))
            if invalid(l):
                raise InvalidLoss()

            newSigns = [p.derivative > 0 for p in parameters]
            for i, p in enumerate(parameters):
                if p.derivative > 0:
                    p.data -= lr[i]
                elif p.derivative < 0:
                    p.data += lr[i]
                if previousSign[i] is not None:
                    if previousSign[i] == newSigns[i]:
                        lr[i] *= grow
                    else:
                        lr[i] *= decay
            previousSign = newSigns

        return self.data


class Placeholder(DN):
    COUNTER = 0

    def __init__(self, initialValue=0., name=None):
        self.data = initialValue
        super(Placeholder, self).__init__([])
        if name is None:
            name = "p_" + str(Placeholder.COUNTER)
            Placeholder.COUNTER += 1
        self.name = name

    @staticmethod
    def named(namePrefix, initialValue=0.):
        p = Placeholder(initialValue, namePrefix + str(Placeholder.COUNTER))
        Placeholder.COUNTER += 1
        return p

    def __str__(self):
        return "Placeholder(%s = %s)" % (self.name, self.data)

    @staticmethod
    def maybe(x):
        if isinstance(x, DN):
            return x
        return Placeholder(float(x))

    def forward(self): return self.data

    def backward(self): return []


class Clamp(DN):
    def __init__(self, x, l, u):
        assert u > l
        self.l = l
        self.u = u
        super(Clamp, self).__init__([x])
        self.name = "clamp"

    def forward(self, x):
        if x > self.u:
            return self.u
        if x < self.l:
            return self.l
        return x

    def backward(self, x):
        if x > self.u or x < self.l:
            return [0.]
        else:
            return [1.]


class Addition(DN):
    def __init__(self, x, y):
        super(Addition, self).__init__([x, y])
        self.name = '+'

    def forward(self, x, y): return x + y

    def backward(self, x, y): return [1., 1.]


class Subtraction(DN):
    def __init__(self, x, y):
        super(Subtraction, self).__init__([x, y])
        self.name = '-'

    def forward(self, x, y): return x - y

    def backward(self, x, y): return [1., -1.]


class Negation(DN):
    def __init__(self, x):
        super(Negation, self).__init__([x])
        self.name = '-'

    def forward(self, x): return -x

    def backward(self, x): return [-1.]


class AbsoluteValue(DN):
    def __init__(self, x):
        super(AbsoluteValue, self).__init__([x])
        self.name = 'abs'

    def forward(self, x): return abs(x)

    def backward(self, x):
        if x > 0:
            return [1.]
        return [-1.]


class Multiplication(DN):
    def __init__(self, x, y):
        super(Multiplication, self).__init__([x, y])
        self.name = '*'

    def forward(self, x, y): return x * y

    def backward(self, x, y): return [y, x]


class Division(DN):
    def __init__(self, x, y):
        super(Division, self).__init__([x, y])
        self.name = '/'

    def forward(self, x, y): return x / y

    def backward(self, x, y): return [1.0 / y, -x / (y * y)]


class Square(DN):
    def __init__(self, x):
        super(Square, self).__init__([x])
        self.name = 'sq'

    def forward(self, x): return x * x

    def backward(self, x): return [2 * x]


class Exponentiation(DN):
    def __init__(self, x):
        super(Exponentiation, self).__init__([x])
        self.name = 'exp'

    def forward(self, x): return math.exp(x)

    def backward(self, x): return [math.exp(x)]


class Logarithm(DN):
    def __init__(self, x):
        super(Logarithm, self).__init__([x])
        self.name = 'log'

    def forward(self, x): return math.log(x)

    def backward(self, x): return [1. / x]


class LSE(DN):
    def __init__(self, xs):
        super(LSE, self).__init__(xs)
        self.name = 'LSE'

    def forward(self, *xs):
        m = max(xs)
        return m + math.log(sum(math.exp(y - m) for y in xs))

    def backward(self, *xs):
        m = max(xs)
        zm = sum(math.exp(x - m) for x in xs)
        return [math.exp(x - m) / zm for x in xs]


if __name__ == "__main__":
    x = Placeholder(10., "x")
    y = Placeholder(2., "y")
    z = x - LSE([x, y])
    z.updateNetwork()
    eprint("dL/dx = %f\tdL/dy = %f" % (x.derivative, y.derivative))

    x.data = 2.
    y.data = 10.
    z.updateNetwork()
    eprint("dL/dx = %f\tdL/dy = %f" % (x.differentiate(), y.differentiate()))

    x.data = 2.
    y.data = 2.
    z.updateNetwork()
    eprint("z = ", z.data, z)
    eprint("dL/dx = %f\tdL/dy = %f" % (x.differentiate(), y.differentiate()))

    loss = -z
    eprint(loss)

    lr = 0.001
    loss.gradientDescent([x, y], steps=10000, update=1000)
